File: backend/app/core/security.py
"""
生产环境安全配置
实现全面的安全措施和DSGVO合规要求
"""

import time
import hashlib
import secrets
from typing import Dict, List, Optional, Set
from datetime import datetime, timedelta
from fastapi import Request, Response, HTTPException, Depends
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import JSONResponse
import logging

logger = logging.getLogger(__name__)

# 安全配置
class SecurityConfig:
    """安全配置类"""
    
    # Content Security Policy配置
    CSP_POLICY = {
        'default-src': ["'self'"],
        'script-src': [
            "'self'",
            "'unsafe-inline'",  # 仅用于内联脚本，生产环境应移除
            "https://www.googletagmanager.com",
            "https://www.google-analytics.com",
            "https://cdn.jsdelivr.net"
        ],
        'style-src': [
            "'self'",
            "'unsafe-inline'",
            "https://fonts.googleapis.com",
            "https://cdn.jsdelivr.net"
        ],
        'font-src': [
            "'self'",
            "https://fonts.gstatic.com"
        ],
        'img-src': [
            "'self'",
            "data:",
            "https:",
            "blob:"
        ],
        'connect-src': [
            "'self'",
            "https://api.zinses-rechner.de",
            "https://www.google-analytics.com"
        ],
        'frame-ancestors': ["'none'"],
        'base-uri': ["'self'"],
        'form-action': ["'self'"],
        'upgrade-insecure-requests': []
    }
    
    # 安全头配置
    SECURITY_HEADERS = {
        'X-Frame-Options': 'DENY',
        'X-Content-Type-Options': 'nosniff',
        'X-XSS-Protection': '1; mode=block',
        'Referrer-Policy': 'strict-origin-when-cross-origin',
        'Permissions-Policy': 'camera=(), microphone=(), geolocation=(), payment=()',
        'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload',
        'Cross-Origin-Embedder-Policy': 'require-corp',
        'Cross-Origin-Opener-Policy': 'same-origin',
        'Cross-Origin-Resource-Policy': 'same-origin'
    }
    
    # CORS配置
    CORS_CONFIG = {
        'allowed_origins': [
            'https://zinses-rechner.de',
            'https://www.zinses-rechner.de',
            'https://staging.zinses-rechner.de'
        ],
        'allowed_methods': ['GET', 'POST', 'OPTIONS'],
        'allowed_headers': ['Content-Type', 'Authorization', 'X-Requested-With'],
        'max_age': 86400  # 24小时
    }
    
    # 请求限流配置
    RATE_LIMIT_CONFIG = {
        'requests_per_minute': 60,
        'requests_per_hour': 1000,
        'requests_per_day': 10000,
        'burst_limit': 10,
        'window_size': 60  # 秒
    }


class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """安全头中间件"""
    
    def __init__(self, app, config: SecurityConfig = None):
        super().__init__(app)
        self.config = config or SecurityConfig()
    
    def generate_csp_header(self) -> str:
        """生成CSP头"""
        csp_parts = []
        for directive, sources in self.config.CSP_POLICY.items():
            if sources:
                csp_parts.append(f"{directive} {' '.join(sources)}")
            else:
                csp_parts.append(directive)
        return '; '.join(csp_parts)
    
    async def dispatch(self, request: Request, call_next):
        response = await call_next(request)
        
        # 添加安全头
        for header, value in self.config.SECURITY_HEADERS.items():
            response.headers[header] = value
        
        # 添加CSP头
        response.headers['Content-Security-Policy'] = self.generate_csp_header()
        
        # 添加DSGVO合规头
        response.headers['X-Robots-Tag'] = 'index, follow'
        response.headers['X-Permitted-Cross-Domain-Policies'] = 'none'
        
        return response


class RateLimitMiddleware(BaseHTTPMiddleware):
    """请求限流中间件"""
    
    def __init__(self, app, config: SecurityConfig = None):
        super().__init__(app)
        self.config = config or SecurityConfig()
        self.request_counts: Dict[str, List[float]] = {}
        self.blocked_ips: Set[str] = set()
        self.whitelist_ips: Set[str] = {
            '127.0.0.1',
            '::1',
            'localhost'
        }
    
    def get_client_ip(self, request: Request) -> str:
        """获取客户端IP地址"""
        # 检查Cloudflare头
        if 'CF-Connecting-IP' in request.headers:
            return request.headers['CF-Connecting-IP']
        
        # 检查代理头
        if 'X-Forwarded-For' in request.headers:
            return request.headers['X-Forwarded-For'].split(',')[0].strip()
        
        if 'X-Real-IP' in request.headers:
            return request.headers['X-Real-IP']
        
        # 回退到直接连接IP
        return request.client.host if request.client else 'unknown'
    
    def is_rate_limited(self, ip: str) -> bool:
        """检查IP是否被限流"""
        if ip in self.whitelist_ips:
            return False
        
        if ip in self.blocked_ips:
            return True
        
        current_time = time.time()
        window_start = current_time - self.config.RATE_LIMIT_CONFIG['window_size']
        
        # 清理过期记录
        if ip in self.request_counts:
            self.request_counts[ip] = [
                timestamp for timestamp in self.request_counts[ip]
                if timestamp > window_start
            ]
        else:
            self.request_counts[ip] = []
        
        # 检查请求频率
        request_count = len(self.request_counts[ip])
        limit = self.config.RATE_LIMIT_CONFIG['requests_per_minute']
        
        if request_count >= limit:
            # 临时阻止IP
            self.blocked_ips.add(ip)
            logger.warning(f"IP {ip} rate limited: {request_count} requests in {self.config.RATE_LIMIT_CONFIG['window_size']}s")
            return True
        
        # 记录请求
        self.request_counts[ip].append(current_time)
        return False
    
    async def dispatch(self, request: Request, call_next):
        client_ip = self.get_client_ip(request)
        
        # 检查是否被限流
        if self.is_rate_limited(client_ip):
            return JSONResponse(
                status_code=429,
                content={
                    'error': 'RATE_LIMIT_EXCEEDED',
                    'message': 'Zu viele Anfragen. Bitte versuchen Sie es später erneut.',
                    'retry_after': self.config.RATE_LIMIT_CONFIG['window_size'],
                    'timestamp': datetime.now().isoformat()
                },
                headers={
                    'Retry-After': str(self.config.RATE_LIMIT_CONFIG['window_size']),
                    'X-RateLimit-Limit': str(self.config.RATE_LIMIT_CONFIG['requests_per_minute']),
                    'X-RateLimit-Remaining': '0',
                    'X-RateLimit-Reset': str(int(time.time() + self.config.RATE_LIMIT_CONFIG['window_size']))
                }
            )
        
        # 添加限流头
        remaining = self.config.RATE_LIMIT_CONFIG['requests_per_minute'] - len(self.request_counts.get(client_ip, []))
        response = await call_next(request)
        
        response.headers['X-RateLimit-Limit'] = str(self.config.RATE_LIMIT_CONFIG['requests_per_minute'])
        response.headers['X-RateLimit-Remaining'] = str(max(0, remaining - 1))
        response.headers['X-RateLimit-Reset'] = str(int(time.time() + self.config.RATE_LIMIT_CONFIG['window_size']))
        
        return response


class APIKeyAuth:
    """API密钥认证"""
    
    def __init__(self):
        self.api_keys: Dict[str, Dict[str, any]] = {
            # 示例API密钥配置（实际应从环境变量读取）
            # 'zr_prod_key_123': {
            #     'name': 'Production Key',
            #     'permissions': ['calculate', 'export'],
            #     'rate_limit': 1000,
            #     'expires_at': None
            # }
        }
        self.security = HTTPBearer(auto_error=False)
    
    async def verify_api_key(self, credentials: Optional[HTTPAuthorizationCredentials] = Depends(HTTPBearer(auto_error=False))) -> Optional[Dict[str, any]]:
        """验证API密钥"""
        if not credentials:
            return None
        
        api_key = credentials.credentials
        key_info = self.api_keys.get(api_key)
        
        if not key_info:
            raise HTTPException(
                status_code=401,
                detail={
                    'error': 'INVALID_API_KEY',
                    'message': 'Ungültiger API-Schlüssel'
                }
            )
        
        # 检查过期时间
        if key_info.get('expires_at') and datetime.now() > key_info['expires_at']:
            raise HTTPException(
                status_code=401,
                detail={
                    'error': 'API_KEY_EXPIRED',
                    'message': 'API-Schlüssel ist abgelaufen'
                }
            )
        
        return key_info


class InputSanitizer:
    """输入数据清理和验证"""
    
    @staticmethod
    def sanitize_string(value: str, max_length: int = 1000) -> str:
        """清理字符串输入"""
        if not isinstance(value, str):
            raise ValueError("Input must be a string")
        
        # 移除危险字符
        dangerous_chars = ['<', '>', '"', "'", '&', '\x00', '\n', '\r', '\t']
        for char in dangerous_chars:
            value = value.replace(char, '')
        
        # 限制长度
        if len(value) > max_length:
            value = value[:max_length]
        
        return value.strip()
    
    @staticmethod
    def sanitize_number(value: any, min_val: float = None, max_val: float = None) -> float:
        """清理数字输入"""
        try:
            num_value = float(value)
            
            # 检查范围
            if min_val is not None and num_value < min_val:
                raise ValueError(f"Value must be >= {min_val}")
            
            if max_val is not None and num_value > max_val:
                raise ValueError(f"Value must be <= {max_val}")
            
            # 检查特殊值
            if not isinstance(num_value, (int, float)) or num_value != num_value:  # NaN检查
                raise ValueError("Invalid number")
            
            return num_value
            
        except (ValueError, TypeError) as e:
            raise ValueError(f"Invalid number format: {e}")
    
    @staticmethod
    def validate_calculation_input(data: Dict[str, any]) -> Dict[str, any]:
        """验证计算输入数据"""
        sanitized = {}
        
        # 验证本金
        sanitized['principal'] = InputSanitizer.sanitize_number(
            data.get('principal'), 
            min_val=1, 
            max_val=10_000_000
        )
        
        # 验证年利率
        sanitized['annual_rate'] = InputSanitizer.sanitize_number(
            data.get('annual_rate'), 
            min_val=0.01, 
            max_val=20
        )
        
        # 验证年限
        sanitized['years'] = int(InputSanitizer.sanitize_number(
            data.get('years'), 
            min_val=1, 
            max_val=50
        ))
        
        # 验证月供
        sanitized['monthly_payment'] = InputSanitizer.sanitize_number(
            data.get('monthly_payment', 0), 
            min_val=0, 
            max_val=50_000
        )
        
        # 验证复利频率
        compound_frequency = data.get('compound_frequency', 'monthly')
        allowed_frequencies = ['monthly', 'quarterly', 'yearly']
        if compound_frequency not in allowed_frequencies:
            raise ValueError(f"Invalid compound frequency. Allowed: {allowed_frequencies}")
        sanitized['compound_frequency'] = compound_frequency
        
        return sanitized


class SecurityAuditLogger:
    """安全审计日志"""
    
    def __init__(self):
        self.security_events: List[Dict[str, any]] = []
        self.max_events = 10000
    
    def log_security_event(
        self, 
        event_type: str, 
        ip: str, 
        user_agent: str = None,
        details: Dict[str, any] = None
    ):
        """记录安全事件"""
        event = {
            'timestamp': datetime.now().isoformat(),
            'event_type': event_type,
            'ip': ip,
            'user_agent': user_agent,
            'details': details or {},
            'event_id': secrets.token_hex(16)
        }
        
        self.security_events.append(event)
        
        # 限制事件数量
        if len(self.security_events) > self.max_events:
            self.security_events = self.security_events[-self.max_events//2:]
        
        # 记录到日志
        logger.warning(f"Security Event: {event_type} from {ip}", extra=event)
    
    def get_security_events(self, event_type: str = None, hours: int = 24) -> List[Dict[str, any]]:
        """获取安全事件"""
        cutoff_time = datetime.now() - timedelta(hours=hours)
        
        filtered_events = [
            event for event in self.security_events
            if datetime.fromisoformat(event['timestamp']) > cutoff_time
        ]
        
        if event_type:
            filtered_events = [
                event for event in filtered_events
                if event['event_type'] == event_type
            ]
        
        return filtered_events


class DDoSProtection:
    """DDoS保护"""
    
    def __init__(self):
        self.suspicious_ips: Dict[str, Dict[str, any]] = {}
        self.blocked_ips: Set[str] = set()
        self.request_patterns: Dict[str, List[float]] = {}
    
    def analyze_request_pattern(self, ip: str, request: Request) -> bool:
        """分析请求模式，检测可疑行为"""
        current_time = time.time()
        
        # 初始化IP记录
        if ip not in self.request_patterns:
            self.request_patterns[ip] = []
        
        # 添加当前请求时间
        self.request_patterns[ip].append(current_time)
        
        # 清理1分钟前的记录
        self.request_patterns[ip] = [
            timestamp for timestamp in self.request_patterns[ip]
            if current_time - timestamp <= 60
        ]
        
        # 检测异常模式
        recent_requests = len(self.request_patterns[ip])
        
        # 检测高频请求
        if recent_requests > 100:  # 1分钟内超过100次请求
            self.mark_suspicious(ip, 'HIGH_FREQUENCY', {
                'requests_per_minute': recent_requests,
                'user_agent': request.headers.get('User-Agent', 'unknown')
            })
            return True
        
        # 检测相同请求重复
        if recent_requests > 20:
            # 简化的重复检测（实际应该检查请求内容）
            if self.is_repetitive_pattern(self.request_patterns[ip]):
                self.mark_suspicious(ip, 'REPETITIVE_PATTERN', {
                    'pattern_detected': True,
                    'requests_count': recent_requests
                })
                return True
        
        return False
    
    def is_repetitive_pattern(self, timestamps: List[float]) -> bool:
        """检测重复模式"""
        if len(timestamps) < 10:
            return False
        
        # 检查请求间隔是否过于规律
        intervals = [timestamps[i] - timestamps[i-1] for i in range(1, len(timestamps))]
        avg_interval = sum(intervals) / len(intervals)
        
        # 如果间隔过于规律（标准差很小），可能是机器人
        variance = sum((interval - avg_interval) ** 2 for interval in intervals) / len(intervals)
        std_dev = variance ** 0.5
        
        return std_dev < 0.1 and avg_interval < 1.0  # 间隔小于1秒且非常规律
    
    def mark_suspicious(self, ip: str, reason: str, details: Dict[str, any]):
        """标记可疑IP"""
        if ip not in self.suspicious_ips:
            self.suspicious_ips[ip] = {
                'first_seen': time.time(),
                'reasons': [],
                'request_count': 0,
                'blocked': False
            }
        
        self.suspicious_ips[ip]['reasons'].append({
            'reason': reason,
            'timestamp': time.time(),
            'details': details
        })
        
        # 如果多次标记为可疑，则阻止
        if len(self.suspicious_ips[ip]['reasons']) >= 3:
            self.blocked_ips.add(ip)
            self.suspicious_ips[ip]['blocked'] = True
            logger.critical(f"IP {ip} blocked due to suspicious activity: {reason}")
    
    def is_blocked(self, ip: str) -> bool:
        """检查IP是否被阻止"""
        return ip in self.blocked_ips


# 全局实例
security_audit_logger = SecurityAuditLogger()
ddos_protection = DDoSProtection()
api_key_auth = APIKeyAuth()


def get_security_audit_logger() -> SecurityAuditLogger:
    """获取安全审计日志实例"""
    return security_audit_logger


def get_ddos_protection() -> DDoSProtection:
    """获取DDoS保护实例"""
    return ddos_protection


def get_api_key_auth() -> APIKeyAuth:
    """获取API密钥认证实例"""
    return api_key_auth


async def security_middleware(request: Request, call_next):
    """综合安全中间件"""
    client_ip = ddos_protection.get_client_ip(request) if hasattr(ddos_protection, 'get_client_ip') else request.client.host
    
    # DDoS保护检查
    if ddos_protection.is_blocked(client_ip):
        security_audit_logger.log_security_event(
            'BLOCKED_REQUEST',
            client_ip,
            request.headers.get('User-Agent'),
            {'reason': 'IP_BLOCKED', 'path': request.url.path}
        )
        
        return JSONResponse(
            status_code=403,
            content={
                'error': 'ACCESS_DENIED',
                'message': 'Zugriff verweigert. Kontaktieren Sie den Support, wenn Sie glauben, dass dies ein Fehler ist.',
                'timestamp': datetime.now().isoformat()
            }
        )
    
    # 分析请求模式
    if ddos_protection.analyze_request_pattern(client_ip, request):
        security_audit_logger.log_security_event(
            'SUSPICIOUS_PATTERN',
            client_ip,
            request.headers.get('User-Agent'),
            {'path': request.url.path, 'method': request.method}
        )
    
    # 执行请求
    response = await call_next(request)
    
    # 添加安全响应头
    response.headers['X-Request-ID'] = secrets.token_hex(16)
    response.headers['X-Security-Scan'] = 'passed'
    
    return response

File: cloudflare-workers/api-worker/src/services/security.ts
/**
 * 安全服务 - Cloudflare Workers版本
 * 实现请求验证、速率限制和安全头配置
 */

import { Env } from '../index'

export interface SecurityCheckResult {
  valid: boolean
  message?: string
  code?: string
}

export interface RateLimitResult {
  allowed: boolean
  retryAfter: number
  remaining: number
}

export class SecurityService {
  private env: Env

  constructor(env: Env) {
    this.env = env
  }

  /**
   * 验证请求安全性
   */
  async validateRequest(request: Request, env: Env): Promise<SecurityCheckResult> {
    try {
      // 检查请求方法
      if (!['GET', 'POST', 'OPTIONS'].includes(request.method)) {
        return {
          valid: false,
          message: 'Ungültige HTTP-Methode',
          code: 'INVALID_METHOD'
        }
      }

      // 检查Content-Type（对于POST请求）
      if (request.method === 'POST') {
        const contentType = request.headers.get('Content-Type')
        if (!contentType || !contentType.includes('application/json')) {
          return {
            valid: false,
            message: 'Content-Type muss application/json sein',
            code: 'INVALID_CONTENT_TYPE'
          }
        }
      }

      // 检查User-Agent（防止机器人）
      const userAgent = request.headers.get('User-Agent')
      if (!userAgent || this.isSuspiciousUserAgent(userAgent)) {
        return {
          valid: false,
          message: 'Verdächtiger User-Agent erkannt',
          code: 'SUSPICIOUS_USER_AGENT'
        }
      }

      // 检查Referer（防止直接API调用）
      if (env.ENVIRONMENT === 'production') {
        const referer = request.headers.get('Referer')
        const origin = request.headers.get('Origin')
        
        if (!referer && !origin) {
          // 允许直接API调用，但记录
          console.warn('Direct API call detected (no referer/origin)')
        } else if (referer && !this.isValidReferer(referer, env.CORS_ORIGIN)) {
          return {
            valid: false,
            message: 'Ungültiger Referer',
            code: 'INVALID_REFERER'
          }
        }
      }

      // 检查请求大小
      const contentLength = request.headers.get('Content-Length')
      if (contentLength && parseInt(contentLength) > 10240) { // 10KB限制
        return {
          valid: false,
          message: 'Anfrage zu groß',
          code: 'REQUEST_TOO_LARGE'
        }
      }

      return { valid: true }

    } catch (error) {
      console.error('Security validation error:', error)
      return {
        valid: false,
        message: 'Sicherheitsprüfung fehlgeschlagen',
        code: 'SECURITY_CHECK_FAILED'
      }
    }
  }

  /**
   * 速率限制检查
   */
  async checkRateLimit(request: Request, env: Env): Promise<RateLimitResult> {
    try {
      const clientIP = request.headers.get('CF-Connecting-IP') || 'unknown'
      const rateLimitKey = `rate_limit:${clientIP}`
      
      const maxRequests = parseInt(env.RATE_LIMIT_REQUESTS)
      const windowSeconds = parseInt(env.RATE_LIMIT_WINDOW)
      
      // 获取当前计数
      const currentCount = await this.getRateLimitCount(rateLimitKey, windowSeconds)
      
      if (currentCount >= maxRequests) {
        return {
          allowed: false,
          retryAfter: windowSeconds,
          remaining: 0
        }
      }

      // 增加计数
      await this.incrementRateLimitCount(rateLimitKey, windowSeconds)

      return {
        allowed: true,
        retryAfter: 0,
        remaining: maxRequests - currentCount - 1
      }

    } catch (error) {
      console.error('Rate limit check error:', error)
      // 错误时允许请求（fail open）
      return {
        allowed: true,
        retryAfter: 0,
        remaining: 100
      }
    }
  }

  /**
   * 获取CORS头
   */
  getCorsHeaders(allowedOrigin: string): Record<string, string> {
    return {
      'Access-Control-Allow-Origin': allowedOrigin,
      'Access-Control-Allow-Methods': 'GET, POST, OPTIONS',
      'Access-Control-Allow-Headers': 'Content-Type, Authorization, X-Requested-With',
      'Access-Control-Max-Age': '86400',
      'Access-Control-Allow-Credentials': 'false'
    }
  }

  /**
   * 获取安全头
   */
  getSecurityHeaders(): Record<string, string> {
    return {
      'X-Content-Type-Options': 'nosniff',
      'X-Frame-Options': 'DENY',
      'X-XSS-Protection': '1; mode=block',
      'Referrer-Policy': 'strict-origin-when-cross-origin',
      'Content-Security-Policy': this.getCSPHeader(),
      'Strict-Transport-Security': 'max-age=31536000; includeSubDomains',
      'Permissions-Policy': 'camera=(), microphone=(), geolocation=()'
    }
  }

  /**
   * 生成CSP头
   */
  private getCSPHeader(): string {
    const csp = [
      "default-src 'self'",
      "script-src 'self' 'unsafe-inline' https://www.googletagmanager.com",
      "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
      "font-src 'self' https://fonts.gstatic.com",
      "img-src 'self' data: https:",
      "connect-src 'self' https://api.zinses-rechner.de",
      "frame-ancestors 'none'",
      "base-uri 'self'",
      "form-action 'self'"
    ]
    
    return csp.join('; ')
  }

  /**
   * 检查可疑User-Agent
   */
  private isSuspiciousUserAgent(userAgent: string): boolean {
    const suspiciousPatterns = [
      /bot/i,
      /crawler/i,
      /spider/i,
      /scraper/i,
      /curl/i,
      /wget/i,
      /python/i,
      /java/i,
      /go-http/i
    ]

    // 允许的机器人
    const allowedBots = [
      /googlebot/i,
      /bingbot/i,
      /facebookexternalhit/i,
      /twitterbot/i,
      /linkedinbot/i
    ]

    // 检查是否是允许的机器人
    if (allowedBots.some(pattern => pattern.test(userAgent))) {
      return false
    }

    // 检查是否是可疑的User-Agent
    return suspiciousPatterns.some(pattern => pattern.test(userAgent))
  }

  /**
   * 验证Referer
   */
  private isValidReferer(referer: string, allowedOrigin: string): boolean {
    try {
      const refererUrl = new URL(referer)
      const allowedUrl = new URL(allowedOrigin)
      
      return refererUrl.hostname === allowedUrl.hostname
    } catch {
      return false
    }
  }

  /**
   * 获取速率限制计数
   */
  private async getRateLimitCount(key: string, windowSeconds: number): Promise<number> {
    try {
      const data = await env.CACHE.get(key, 'json') as any
      
      if (!data) {
        return 0
      }

      // 检查时间窗口
      const now = Date.now()
      const windowStart = now - (windowSeconds * 1000)
      
      if (data.timestamp < windowStart) {
        // 时间窗口过期，重置计数
        await env.CACHE.delete(key)
        return 0
      }

      return data.count || 0
      
    } catch (error) {
      console.error('Failed to get rate limit count:', error)
      return 0
    }
  }

  /**
   * 增加速率限制计数
   */
  private async incrementRateLimitCount(key: string, windowSeconds: number): Promise<void> {
    try {
      const now = Date.now()
      const data = await env.CACHE.get(key, 'json') as any || { count: 0, timestamp: now }
      
      data.count += 1
      data.timestamp = now
      
      await env.CACHE.put(key, JSON.stringify(data), {
        expirationTtl: windowSeconds
      })
      
    } catch (error) {
      console.error('Failed to increment rate limit count:', error)
    }
  }

  /**
   * 生成安全的会话ID
   */
  generateSessionId(): string {
    const array = new Uint8Array(16)
    crypto.getRandomValues(array)
    return Array.from(array, byte => byte.toString(16).padStart(2, '0')).join('')
  }

  /**
   * 验证请求签名（可选的高级安全功能）
   */
  async validateRequestSignature(request: Request, secret: string): Promise<boolean> {
    try {
      const signature = request.headers.get('X-Signature')
      if (!signature) {
        return false
      }

      const body = await request.text()
      const expectedSignature = await this.generateSignature(body, secret)
      
      return signature === expectedSignature
      
    } catch (error) {
      console.error('Signature validation error:', error)
      return false
    }
  }

  /**
   * 生成请求签名
   */
  private async generateSignature(data: string, secret: string): Promise<string> {
    const encoder = new TextEncoder()
    const key = await crypto.subtle.importKey(
      'raw',
      encoder.encode(secret),
      { name: 'HMAC', hash: 'SHA-256' },
      false,
      ['sign']
    )
    
    const signature = await crypto.subtle.sign('HMAC', key, encoder.encode(data))
    return Array.from(new Uint8Array(signature))
      .map(b => b.toString(16).padStart(2, '0'))
      .join('')
  }
}
